src/promptflow-evals/promptflow/evals/synthetic/adversarial_simulator.py

```diff
@@ -0,0 +1,316 @@
+    ) -> List[Dict]:
+        user_bot = self._setup_bot(role=ConversationRole.USER, template=template, parameters=parameters)
+        system_bot = self._setup_bot(
+            target=target, role=ConversationRole.ASSISTANT, template=template, parameters=parameters
+        )
+        bots = [user_bot, system_bot]
+        asyncHttpClient = AsyncHTTPClientWithRetry(
+            n_retry=api_call_retry_limit,
+            retry_timeout=api_call_retry_sleep_sec,
+            logger=logger,
+        )
+        async with semaphore:
+            async with asyncHttpClient.client as session:
+                _, conversation_history = await simulate_conversation(
+                    bots=bots,
+                    session=session,
+                    turn_limit=max_conversation_turns,
+                    api_call_delay_sec=api_call_delay_sec,
+                )
+        return self._to_chat_protocol(conversation_history=conversation_history, template_parameters=parameters)
+
+    def _get_user_proxy_completion_model(self, template_key, template_parameters):
+        return ProxyChatCompletionsModel(
+            name="raisvc_proxy_model",
+            template_key=template_key,
+            template_parameters=template_parameters,
+            endpoint_url=self.rai_client.simulation_submit_endpoint,
+            token_manager=self.token_manager,
+            api_version="2023-07-01-preview",
+            max_tokens=1200,
+            temperature=0.0,
+        )
+
+    def _setup_bot(self, *, role, template, parameters, target: Callable = None):
+        if role == ConversationRole.USER:
+            model = self._get_user_proxy_completion_model(
+                template_key=template.template_name, template_parameters=parameters
+            )
+            return ConversationBot(
+                role=role,
+                model=model,
+                conversation_template=str(template),
+                instantiation_parameters=parameters,
+            )
+
+        if role == ConversationRole.ASSISTANT:
+            dummy_model = lambda: None  # noqa: E731
+            dummy_model.name = "dummy_model"
+            return CallbackConversationBot(
+                callback=target,
+                role=role,
+                model=dummy_model,
+                user_template=str(template),
+                user_template_parameters=parameters,
+                conversation_template="",
+                instantiation_parameters={},
+            )
+        return ConversationBot(
+            role=role,
+            model=model,
+            conversation_template=template,
+            instantiation_parameters=parameters,
+        )
+
+    def _join_conversation_starter(self, parameters, to_join):
+        key = "conversation_starter"
+        if key in parameters.keys():
+            parameters[key] = f"{to_join} {parameters[key]}"
+        else:
+            parameters[key] = to_join
+
+        return parameters
+
+    def call_sync(

```

Comment:
shall we introduce two class for sync/async scenario? 
AdversarialSimulator & AsyncAdversarialSimulator

Reply:
All the current customers are happy with the async interface. The sync interface in my opinion is not used and we can remove it once we get enough usage on the async.

Reply:
We will build a sync interface if customers ask for it post build.